src/main/scala/edu/twitter/model/impl/neuralnetwork/DataIterator.scala

```diff
@@ -0,0 +1,179 @@
+import org.deeplearning4j.text.tokenization.tokenizerfactory.{DefaultTokenizerFactory, TokenizerFactory}
+import org.nd4j.linalg.dataset.DataSet
+import org.nd4j.linalg.dataset.api.DataSetPreProcessor
+import org.nd4j.linalg.dataset.api.iterator.DataSetIterator
+import org.nd4j.linalg.factory.Nd4j
+import org.nd4j.linalg.indexing.{INDArrayIndex, NDArrayIndex}
+
+import scala.collection.JavaConverters._
+
+
+/**
+  * An Iterator used to convert the training (or the testing) data to the format
+  * that the neural network model understands.
+  *
+  * @param data           The labeled tweets
+  * @param wordVectors    WordVectors object
+  * @param batchSize      Size of each minibatch for training
+  * @param truncateLength If reviews exceed
+  */
+@throws[IOException]
+class DataIterator(val data: RDD[Row],
+                   val wordVectors: WordVectors,
+                   val batchSize: Int,
+                   val truncateLength: Int) extends DataSetIterator {
+
+  private final val vectorSize = wordVectors.getWordVector(wordVectors.vocab.wordAtIndex(0)).length
+
+  private val dataList = data.collect()
+  final private val tokenizerFactory: TokenizerFactory = new DefaultTokenizerFactory
+  tokenizerFactory.setTokenPreProcessor(new CommonPreprocessor)
+
+  private var _cursor: Int = 0
+
+  def next(num: Int): DataSet = {
+    if (_cursor >= dataList.length)
+      throw new NoSuchElementException
+    try {
+      nextDataSet(num)
+    } catch {
+      case e: IOException =>
+        throw new RuntimeException(e)
+    }
+  }
+
+  @throws[IOException]
+  private def nextDataSet(num: Int): DataSet = {
+    val reviews = new util.ArrayList[String]()
+    val positive = new util.ArrayList[Boolean]()
+    var i = 0
+
+    while (i < num && _cursor < totalExamples) {
+
+      val msg = dataList(_cursor).getAs[String]("msg")
+      val label = dataList(_cursor).getAs[Double]("label")
+
+      //TODO(elhawaty): Find why some rows return null.
+      if (msg != null) {
+        reviews.add(msg)
+        positive.add(if (label == 1.0) true else false)
+        i += 1
+      } else {
+        println("Invalid tweet:" + dataList(_cursor))
+      }
+
+      _cursor += 1
+    }
+
+    //Second: tokenize reviews and filter out unknown words
+    val allTokens: util.List[util.List[String]] = new util.ArrayList[util.List[String]](reviews.size)
+    var maxLength: Int = 0
+    import scala.collection.JavaConversions._
+    for (s <- reviews) {
+      val tokens = tokenizerFactory.create(s).getTokens
+      val tokensFiltered = new util.ArrayList[String]
+      for (t <- tokens.asScala) {
+        if (wordVectors.hasWord(t)) tokensFiltered.add(t)
+      }
+      allTokens.add(tokensFiltered)
+      maxLength = Math.max(maxLength, tokensFiltered.size)
+    }
+
+    // Workaround, just in case the word2vec doesn't recognise all the words in the batch which is unlikely to happen as we are using Google word2vec.
+    if (maxLength == 0) {
```

Comment:
I guess so, all you need is one word that you are sure it will be recognized, anyway, this is unlikely to happen and I am planning to remove it just after we train the model with the whole data.